pkg/networkservice/common/heal/client.go

```diff
@@ -70,61 +72,79 @@ func NewClient(client networkservice.MonitorConnectionClient, onHeal *networkser
 		updateExecutor:    serialize.NewExecutor(),
 		eventReceiver:     nil, // This is intentionally nil
 		recvEventExecutor: serialize.NewExecutor(),
+		chainContext:      ctx,
 	}
+
 	if rv.onHeal == nil {
 		rv.onHeal = addressof.NetworkServiceClient(rv)
 	}
-	rv.updateExecutor.AsyncExec(func() {
-		runtime.SetFinalizer(rv, func(f *healClient) {
-			f.updateExecutor.AsyncExec(func() {
-				if f.cancelFunc != nil {
-					f.cancelFunc()
-				}
-			})
-		})
-		ctx, cancelFunc := context.WithCancel(context.Background())
-		rv.cancelFunc = cancelFunc
-		// TODO decide what to do about err here
-		recv, _ := rv.client.MonitorConnections(ctx, nil)
-		rv.eventReceiver = recv
-		rv.recvEventExecutor.AsyncExec(rv.recvEvent)
-	})
+
+	rv.init()
 
 	return rv
 }
 
+func (f *healClient) init() {
+	if f.eventReceiver != nil {
+		select {
+		case <-f.eventReceiver.Context().Done():
+			return
+		default:
+			f.eventReceiver = nil
+		}
+	}
+
+	logrus.Info("Creating new eventReceiver")
+
+	recv, err := f.client.MonitorConnections(f.chainContext, &networkservice.MonitorScopeSelector{}, grpc.WaitForReady(true))
+	if err != nil {
+		f.updateExecutor.AsyncExec(func() {
+			f.init()
+		})
+		return
+	}
+
+	f.eventReceiver = recv
+	f.recvEventExecutor.AsyncExec(f.recvEvent)
+}
+
 func (f *healClient) recvEvent() {
 	select {
 	case <-f.eventReceiver.Context().Done():
-		f.eventReceiver = nil
+		return
 	default:
 		event, err := f.eventReceiver.Recv()
-		if err != nil {
-			event = nil
-		}
 		f.updateExecutor.AsyncExec(func() {
+			if err != nil {
+				for id, r := range f.requestors {
+					r()
+					delete(f.reported, id)
+				}
+				f.init()
+				return
+			}
+
 			switch event.GetType() {
 			case networkservice.ConnectionEventType_INITIAL_STATE_TRANSFER:
 				f.reported = event.GetConnections()
+				if event.GetConnections() == nil {
```

Comment:
Why are we zeroing out reported here rather than taking what is coming from INITIAL_STATE_TRANSFER?

And I don't see how we are correctly handling the re-request case for an INITIAL_STATE_TRANSFER that is missing some of the connections we think we have.

Reply:
If we write empty slice to grpc stream, we receive it like `nil`. AFAIK nil and empty map looks the same for grpc.

Reply:
At least I faced with `panic` here if the first INITIAL_STATE_TRANSFER is empty

Reply:
Oh... I see now :)